src/Agent.Plugins/PipelineArtifact/FileContainerProvider.cs

```diff
@@ -0,0 +1,210 @@
+using Minimatch;
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Net.Http;
+using System.Runtime.InteropServices;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using System.Threading.Tasks.Dataflow;
+
+namespace Agent.Plugins.PipelineArtifact
+{
+    class FileContainerProvider : IArtifactProvider
+    {
+        private FileContainerHttpClient containerClient;
+        private CallbackAppTraceSource tracer;
+
+        public FileContainerProvider(VssConnection connection, CallbackAppTraceSource tracer)
+        {
+            containerClient = connection.GetClient<FileContainerHttpClient>();
+            this.tracer = tracer;
+        }
+        public async Task DownloadSingleArtifactAsync(PipelineArtifactDownloadParameters downloadParameters, BuildArtifact buildArtifact, CancellationToken cancellationToken)
+        {
+            await this.DownloadFileContainerAsync(downloadParameters.ProjectId, buildArtifact, downloadParameters.TargetDirectory, downloadParameters.MinimatchFilters, cancellationToken);
+        }
+
+        public async Task DownloadMultipleArtifactsAsync(PipelineArtifactDownloadParameters downloadParameters, IEnumerable<BuildArtifact> buildArtifacts, CancellationToken cancellationToken)
+        {
+            await this.DownloadFileContainersAsync(downloadParameters.ProjectId, buildArtifacts, downloadParameters.TargetDirectory, downloadParameters.MinimatchFilters, cancellationToken);
+        }
+
+        public async Task DownloadFileContainersAsync(Guid projectId, IEnumerable<BuildArtifact> buildArtifacts, string targetDirectory, string[] minimatchFilters, CancellationToken cancellationToken)
+        {
+            if (buildArtifacts.Count() == 0)
+            {
+                return;
+            }
+            foreach (var buildArtifact in buildArtifacts)
+            {
+                var specificPath = Path.Combine(targetDirectory, buildArtifact.Name);
+                Directory.CreateDirectory(specificPath);
+                await DownloadFileContainerAsync(projectId, buildArtifact, specificPath, minimatchFilters, cancellationToken);
+            }
+        }
+
+        private Tuple<long, string> ParseContainerId(string resourceData)
+        {
+            // Example of resourceData: "#/7029766/artifacttool-alpine-x64-Debug"
+            var segments = resourceData.Split('/');
+
+            long containerId;
+            if (segments.Length < 3)
+            {
+                throw new ArgumentException($"Resource data value '{resourceData}' invalid");
+            }
+            if (segments.Length >= 3 && segments[0] == "#" && long.TryParse(segments[1], out containerId))
+            {
+                int index = resourceData.IndexOf('/', resourceData.IndexOf('/') + 1); //Artifact name can also have '/' - hence get everything after second '/'
+                return new Tuple<long, string>(
```

Comment:
why don't we just join all segments after second, I think it's easier to understand